blog/2021-06-17-activity-timeouts.md

```diff
@@ -0,0 +1,258 @@
+
+## Heartbeat Timeout
+
+For long running activities, we recommend recording heartbeats to create more frequent pingbacks from the Activity Worker to Temporal Server. 
+Then, set a `HeartbeatTimeout` to create a failure when Temporal fails to receive a heartbeat from your activity at the expected frequency.
+
+![image](https://user-images.githubusercontent.com/6764957/122424268-287f4800-cfc1-11eb-8417-f52743347afc.png)
+
+By their nature, Heartbeats must be recorded from Activity code using SDK APIs:
+
+<Tabs
+  defaultValue="go"
+  values={[
+    {label: 'Go', value: 'go'},
+    {label: 'Java', value: 'java'},
+  ]
+}>
+
+<TabItem value="go">
+
+```go
+progress := 0
+for hasWork {
+    // Send heartbeat message to the server.
+    activity.RecordHeartbeat(ctx, progress)
+    // Do some work.
+    ...
+    progress++
+}
+```
+
+</TabItem>
+<TabItem value="java">
+
+```java
+while ((read = inputStream.read(bytes)) != -1) {
+  totalRead += read;
+  f.write(bytes, 0, read);
+  // Let the Server know about the download progress.
+  Activity.getExecutionContext().heartbeat(totalRead);
+}
+```
+
+</TabItem>
+</Tabs>
+
+Setting a `HeartbeatTimeout` allows us to retry activities more quickly (e.g. the next minute after a heartbeat is missed) than the `StartToCloseTimeout`, which is necessarily set to as long as the longest possible activity (e.g. five hours later when we are sure the activity should have been completed).
+
+There are some minor nuances to heartbeats that may be of interest:
+
+- You can freely record heartbeats as often as you want - once a minute, or everytime a loop iterates - the SDKs throttle the heartbeats that get sent back anyway.
+- If a `HeartbeatTimeout` isn't set and the activity tries to record one, nothing will be recorded since that information will never be used.
+
+## ScheduleToStart Timeout
+
+We use the `ScheduleToStartTimeout`  to set a maximum limit that an activity should sit in a queue. We recommend that most users monitor the `ScheduleToStart` latency metric and set alerts for them as a [production scaling](https://docs.temporal.io/docs/server/production-deployment/#faq-autoscaling-workers-based-on-task-queue-load) metric, rather than setting an explicit timeout for it.
+
+![image](https://user-images.githubusercontent.com/6764957/122290279-287e3a00-cf26-11eb-8dd6-3133016a0bd9.png)
+
+As a queue timeout, `ScheduleToStart` is unique in that it **cannot be retried** — all a retry would do is pop the activity right back on to the same queue!
+
+The `ScheduleToStartTimeout` is most useful when you have a concrete plan to reroute activity to a different queue, if a given task queue is not draining in adequate time. You can also reschedule a whole set of other activities or do other compensation logic based on this timeout. This is a powerful feature for building ultra-reliable systems, however most users will not need this since you can horizontally scale the number of workers easily.
```

Comment:
```suggestion
The Schedule-To-Start Timeout is most useful when you have a concrete plan to reroute an Activity Task to a different Task Queue, if a given Task Queue is not draining in adequate time.
You can also reschedule a whole set of other Activity Executions or do other compensation logic based on this timeout.
This is a powerful feature for building ultra-reliable systems, however most users will not need this since you can horizontally scale the number of Workers easily.
```